### ISSUE TITLE
`reportInconsistentOverload` false positive when overload implementation returns narrower type than overload signatures

### ISSUE THREAD

**Issue by DetachHead** [DetachHead/basedpyright#1503]
### Description

Code sample in [basedpyright playground](https://basedpyright.com/?typeCheckingMode=all&reportUnusedParameter=false&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAySMApiAIYA2ANFGAG5mVjkAmAUOwAINMsesSwKMDBgAFEgBcUAM4wQASigBaAHyYUMGQDo93XiGZt2g4aInTNMZerkLd%2BzmZFjJM1PgA%2B9pao1EpBSUANoAjAC6UD7yII46UADEUGTgcVAgJDh4AJIoAMZgKLJI8iRaAPKMRvxAA)

```python
from typing import Literal, overload

@overload
def foo(i: str) -> int: ...
@overload
def foo(i: int) -> str: ...


def foo(i: int | str) -> Literal[1] | str: ... # error: reportInconsistentOverload
```

while this can be easily avoided by changing the return type to `int | str` this comes up often if relying on inferred return types:

```py
from typing import overload

@overload
def foo(i: str) -> int: ...
@overload
def foo(i: int) -> str: ...


def foo(i: int | str): # inferred as `-> Literal[1, ""]`
    return 1 if isinstance(i, str) else "" # error: reportInconsistentOverload
```

**Comment by KotlinIsland** [comment]
> the input signature of the implementation should be [assignable](https://typing.python.org/en/latest/spec/glossary.html#term-assignable) to the input signatures of all overloads, and the return type of all overloads should be assignable to the return type of the implementation.

https://typing.python.org/en/latest/spec/overload.html#implementation-consistency

**Comment by KotlinIsland** [comment]
mypy be like
```py
from typing import overload

@overload
def foo(i: str) -> object: ...
@overload
def foo(i: int) -> object: ...

def foo(i: int | str) -> int | str: ... # Overloaded function implementation cannot produce return type of signature 1
```

# 🤣🤣🤣

**Comment by KotlinIsland** [comment]
omfg

```py
from typing import overload

@overload
def foo(i: str) -> int: ...  # no error
@overload
def foo(i: int) -> str: ...  # no error

def foo(i: int | str) -> object:
    return None
```

**Comment by DetachHead** [comment]
pyright does the same thing wtf

**Comment by KotlinIsland** [comment]
it's in the spec to do this

**Comment by KotlinIsland** [comment]
I think a better solution would be:

1. The impl return type must be assignable to each signature, allowing for union distribution (i.e. impl: `int | str` is assignable to `int` if considering each union member)
2. the impl return type must be a subtype of the union of overload return types

**Comment by DetachHead** [comment]
> it's in the spec to do this

That's insane, how can the spec be so blatantly wrong

**Comment by KotlinIsland** [comment]
- https://github.com/python/mypy/issues/12401
- https://github.com/python/mypy/issues/12434
